File: src/lib/rehearsal.ts
import { detectSections } from "./detectSections";
import type {
  BeatMap,
  LyricLine,
  Passage,
  PlayerSettings,
  PracticeEvent,
  PracticeState,
  SectionOccurrence,
  Song,
} from "../types";

export const defaultSettings: PlayerSettings = {
  leadIn: "none",
  countIn: "click",
  beats: 2,
  versesFirst: true,
  singer: 1,
  mix: false,
  fontSize: 48,
  syncOffset: 0,
};
export const formatTime = (time: number) =>
  `${Math.floor(Math.max(0, time) / 60)}:${Math.floor(Math.max(0, time) % 60)
    .toString()
    .padStart(2, "0")}`;
export const playable = (song: Song) =>
  Boolean(song.originalUrl || (song.vocalsUrl && song.instrumentalUrl));
export const usable = (line: LyricLine) =>
  (line.verified || line.confidence >= 0.55) &&
  line.end > line.start &&
  Number.isFinite(line.start + line.end);

export function getSections(song: Song): SectionOccurrence[] {
  const sections: SectionOccurrence[] = [];
  for (const line of detectSections(song)) {
    const previous = sections.at(-1);
    const id =
      line.sectionId ??
      (previous?.name === line.section ? previous.id : `section_${line.id}`);
    if (!previous || previous.id !== id)
      sections.push({
        id,
        name: line.section,
        kind:
          /chorus/i.test(line.section) && !/pre/i.test(line.section)
            ? "chorus"
            : /verse/i.test(line.section)
              ? "verse"
              : "other",
        lineIds: [],
        start: line.start,
        end: line.end,
      });
    const section = sections.at(-1)!;
    section.lineIds.push(line.id);
    section.end = Math.max(section.end, line.end);
  }
  const seen = new Map<string, number>();
  const names = sections.map((s) => s.name);
  return sections.map((s) => {
    const n = (seen.get(s.name) ?? 0) + 1;
    seen.set(s.name, n);
    return {
      ...s,
      name:
        names.filter((n) => n === s.name).length > 1
          ? `${s.name} ${n}`
          : s.name,
    };
  });
}

export function activeLine(lines: LyricLine[], time: number): number {
  return lines.findIndex(
    (line) => usable(line) && time >= line.start && time < line.end,
  );
}
export function feedbackLine(
  lines: LyricLine[],
  time: number,
  heardFrom = 0,
): number {
  const current = activeLine(lines, time);
  if (current >= 0) {
    const line = lines[current];
    const previous = lines[current - 1];
    if (
      previous &&
      previous.end - Math.max(previous.start, heardFrom) >=
        Math.min(0.5, (previous.end - previous.start) / 2) &&
      usable(previous) &&
      line.start - previous.end < 2 &&
      time - line.start < Math.min(0.8, (line.end - line.start) / 2)
    )
      return current - 1;
    return current;
  }
  for (let index = lines.length - 1; index >= 0; index--) {
    if (
      usable(lines[index]) &&
      time >= lines[index].end &&
      time - lines[index].end <= 2
    )
      return index;
  }
  return -1;
}

export function deriveState(events: PracticeEvent[]): PracticeState {
  const byId = new Map(events.map((e) => [e.id, e]));
  const undone = new Set(
    [...byId.values()]
      .filter((e) => e.judgment === "undo")
      .map((e) => e.supersedes),
  );
  // One judgment per line/pass; a changed judgment supersedes without adding a pass.
  const passes = new Map<string, PracticeEvent>();
  for (const event of [...byId.values()].sort((a, b) =>
    a.sessionId === b.sessionId
      ? a.sequence - b.sequence
      : a.createdAt.localeCompare(b.createdAt),
  )) {
    if (event.judgment !== "undo" && !undone.has(event.id))
      passes.set(`${event.sessionId}:${event.passId}:${event.lineId}`, event);
  }
  const state: PracticeState = {};
  for (const e of [...passes.values()].sort(
    (a, b) => a.createdAt.localeCompare(b.createdAt) || a.sequence - b.sequence,
  )) {
    const old = state[e.lineId];
    state[e.lineId] = {
      judgment: e.judgment as "again" | "got-it",
      comfortablePasses:
        e.judgment === "again" ? 0 : (old?.comfortablePasses ?? 0) + 1,
      lastAt: e.createdAt,
    };
  }
  return state;
}

export function makePassage(
  song: Song,
  first: number,
  last = first,
  leadIn: PlayerSettings["leadIn"] = "previous",
): Passage {
  first = Math.max(0, Math.min(first, song.lines.length - 1));
  last = Math.max(first, Math.min(last, song.lines.length - 1));
  const line = song.lines[first];
  const ending = song.lines[last];
  const previous = song.lines[first - 1];
  const start =
    leadIn === "previous" && previous && usable(previous)
      ? previous.start
      : Math.max(0, line.start - (leadIn === "none" ? 0.08 : 2));
  const section = getSections(song).find((s) => s.lineIds.includes(line.id));
  return {
    id: `${line.id}:${ending.id}`,
    lineIds: song.lines.slice(first, last + 1).map((l) => l.id),
    start,
    end: Math.min(song.duration, ending.end + 0.08),
    targetStart: line.start,
    targetEnd: ending.end,
    label: `${section?.name ?? line.section} · ${first === last ? `line ${first + 1}` : `lines ${first + 1}–${last + 1}`}`,
  };
}

export function phrase(
  song: Song,
  index: number,
  settings: PlayerSettings,
  expand = false,
): Passage {
  const section = getSections(song).find((s) =>
    s.lineIds.includes(song.lines[index]?.id),
  );
  if (expand && section) {
    const a = song.lines.findIndex((l) => l.id === section.lineIds[0]);
    return makePassage(
      song,
      a,
      a + section.lineIds.length - 1,
      settings.leadIn,
    );
  }
  const next = song.lines[index + 1];
  const last =
    next && section?.lineIds.includes(next.id) && usable(next)
      ? index + 1
      : index;
  return makePassage(song, index, last, settings.leadIn);
}

export function chooseNext(
  song: Song,
  state: PracticeState,
  history: string[],
  scope: string,
  versesFirst: boolean,
): number | null {
  const sections = getSections(song);
  const scoped = sections.find((s) => s.id === scope);
  const candidates = song.lines
    .map((line, index) => ({
      line,
      index,
      section: sections.find((s) => s.lineIds.includes(line.id)),
    }))
    .filter(
      ({ line }) =>
        usable(line) &&
        line.end <= song.duration &&
        (!scoped || scoped.lineIds.includes(line.id)),
    )
    .filter(({ line }) => !history.slice(-1).includes(line.id));
  if (!candidates.length) return null;
  const normalized = (text: string) =>
    text.toLowerCase().replace(/[^\p{L}\p{N}]/gu, "");
  const comfortableTexts = new Set(
    song.lines
      .filter((l) => state[l.id]?.judgment === "got-it")
      .map((l) => normalized(l.text)),
  );
  const scored = candidates
    .map((c) => {
      const mark = state[c.line.id];
      const visited = history.filter((id) => id === c.line.id).length;
      let score =
        mark?.judgment === "again"
          ? 100
          : mark?.judgment === "got-it"
            ? -30
            : 20;
      score -= visited * 35;
      if (versesFirst && c.section?.kind === "verse") score += 8;
      if (
        versesFirst &&
        c.section?.kind === "chorus" &&
        comfortableTexts.has(normalized(c.line.text)) &&
        mark?.judgment !== "again"
      )
        score -= 25;
      return { ...c, score };
    })
    .sort((a, b) => b.score - a.score || a.index - b.index);
  return scored[0].score > -40 ? scored[0].index : null;
}

export function countPlan(
  map: BeatMap | null,
  start: number,
  settings: PlayerSettings,
): { duration: number; clicks: number[] } {
  if (!map?.verified || settings.countIn === "off" || !map.beats.length)
    return { duration: 0, clicks: [] };
  let index = map.beats.findIndex((t) => t >= start);
  if (index < 0) index = map.beats.length - 1;
  const step =
    index > 0 ? map.beats[index] - map.beats[index - 1] : 60 / map.bpm;
  if (!Number.isFinite(step) || step <= 0) return { duration: 0, clicks: [] };
  // Count into the source entry's fractional beat phase, preserving pickups.
  const previousBeat =
    map.beats[index] > start ? map.beats[index] - step : map.beats[index];
  const phase = Math.max(0, Math.min(step, start - previousBeat));
  const duration = settings.beats * step + phase;
  return {
    duration,
    clicks:
      settings.countIn === "click"
        ? Array.from({ length: settings.beats }, (_, i) => i * step)
        : [],
  };
}

File: services/audio_worker/rehearsal.py
"""Playback-first persistence. Old cards/reviews remain archival only."""
from __future__ import annotations

import json
import math
from collections import Counter
from datetime import datetime, timezone
from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from . import database as db
from .lyrics import parse_lyrics

router = APIRouter(prefix="/api")


def migrate() -> None:
    with db.connect() as connection:
        version = connection.execute("PRAGMA user_version").fetchone()[0]
        if version < 2:
            backup = db.DB_PATH.with_suffix(".v1-backup.sqlite3")
            if not backup.exists():
                import sqlite3
                with sqlite3.connect(backup) as destination:
                    connection.backup(destination)
        connection.executescript("""
            CREATE TABLE IF NOT EXISTS practice_events (
                id TEXT PRIMARY KEY, song_id TEXT NOT NULL, session_id TEXT NOT NULL,
                sequence INTEGER NOT NULL, document TEXT NOT NULL,
                UNIQUE(session_id, sequence)
            );
            CREATE INDEX IF NOT EXISTS idx_practice_song ON practice_events(song_id);
            CREATE TABLE IF NOT EXISTS practice_sessions (
                id TEXT PRIMARY KEY, song_id TEXT NOT NULL, document TEXT NOT NULL, updated_at TEXT NOT NULL
            );
            CREATE TABLE IF NOT EXISTS beat_maps (song_id TEXT PRIMARY KEY, document TEXT NOT NULL);
        """)
        for row in connection.execute("SELECT id, document FROM songs").fetchall():
            song = json.loads(row["document"])
            if song.get("schemaVersion", 0) >= 2:
                continue
            parsed = parse_lyrics(song.get("lyrics", ""))
            occurrence = 0
            previous = None
            for index, line in enumerate(song.get("lines", [])):
                if len(parsed) == len(song["lines"]):
                    occurrence = parsed[index].occurrence
                elif previous != line.get("section"):
                    occurrence += 1
                previous = line.get("section")
                line.setdefault("sectionId", f"{song['id']}_section_{occurrence}")
            song.update(schemaVersion=2, alignmentRevision=song.get("alignmentRevision", 0))
            connection.execute("UPDATE songs SET document=? WHERE id=?", (json.dumps(song), row["id"]))
        connection.execute("PRAGMA user_version=2")


def require_song(song_id: str) -> tuple[dict, dict]:
    found = db.get_song(song_id)
    if not found:
        raise HTTPException(404, "Song not found")
    return found


def sections(song: dict) -> list[dict]:
    result = []
    for line in song.get("lines", []):
        identity = line.get("sectionId", f"section_{line['id']}")
        if not result or result[-1]["id"] != identity:
            name = line.get("section", "Song")
            kind = "chorus" if "chorus" in name.lower() and "pre" not in name.lower() else "verse" if "verse" in name.lower() else "other"
            result.append(dict(id=identity, name=name, kind=kind, lineIds=[], start=line["start"], end=line["end"]))
        result[-1]["lineIds"].append(line["id"])
        result[-1]["end"] = max(result[-1]["end"], line["end"])
    counts = Counter(item["name"] for item in result)
    seen: Counter = Counter()
    for item in result:
        name = item["name"]
        seen[name] += 1
        if counts[name] > 1:
            item["name"] = f"{name} {seen[name]}"
    return result


def public_song(song: dict) -> dict:
    return {**song, "capabilities": {
        "playable": bool(song.get("originalUrl") or (song.get("vocalsUrl") and song.get("instrumentalUrl"))),
        "stems": bool(song.get("vocalsUrl") and song.get("instrumentalUrl")),
        "lineTiming": any(line.get("verified") or line.get("confidence", 0) >= .55 for line in song.get("lines", [])),
    }}


@router.get("/songs/{song_id}/structure")
def get_structure(song_id: str) -> list[dict]:
    return sections(require_song(song_id)[0])


class StructureChange(BaseModel):
    sectionId: str
    name: str = Field(min_length=1, max_length=100)
    revision: int
    splitAt: str | None = None


@router.put("/songs/{song_id}/structure")
def change_structure(song_id: str, change: StructureChange) -> dict:
    song, paths = require_song(song_id)
    if song.get("alignmentRevision", 0) != change.revision:
        raise HTTPException(409, "Song changed. Reload before editing.")
    group = [line for line in song["lines"] if line.get("sectionId") == change.sectionId]
    if not group or (change.splitAt and change.splitAt not in {line["id"] for line in group}):
        raise HTTPException(422, "Unknown section or split line")
    active = not change.splitAt
    for line in group:
        active = active or line["id"] == change.splitAt
        if active:
            line["section"] = change.name.strip()
            if change.splitAt:
                line["sectionId"] = f"section_{change.splitAt}"
    song["alignmentRevision"] = change.revision + 1
    db.save_song(song, paths, datetime.now(timezone.utc).isoformat(), expected_revision=change.revision)
    return public_song(song)


class PracticeEvent(BaseModel):
    id: str = Field(max_length=100)
    songId: str
    sessionId: str = Field(max_length=100)
    sequence: int = Field(ge=0)
    passId: str = Field(max_length=100)
    lineId: str
    judgment: Literal["again", "got-it", "undo"]
    sourceTime: float = Field(ge=0, allow_inf_nan=False)
    createdAt: str
    supersedes: str | None = None


class EventBatch(BaseModel):
    events: list[PracticeEvent] = Field(max_length=200)


@router.post("/practice-events/batch")
def save_events(batch: EventBatch) -> dict:
    # Validation and insert share the same transaction; event IDs make retries safe.
    with db.connect() as connection:
        for event in batch.events:
            row = connection.execute("SELECT document FROM songs WHERE id=?", (event.songId,)).fetchone()
            if not row or event.lineId not in {line["id"] for line in json.loads(row[0])["lines"]}:
                raise HTTPException(422, "Feedback line does not belong to this song")
            existing = connection.execute("SELECT document FROM practice_events WHERE id=?", (event.id,)).fetchone()
            payload = event.model_dump(exclude_none=True)
            if existing:
                if json.loads(existing[0]) != payload:
                    raise HTTPException(409, "Event ID already used for different feedback")
                continue
            owner = connection.execute("SELECT song_id FROM practice_events WHERE session_id=? LIMIT 1", (event.sessionId,)).fetchone()
            if owner and owner[0] != event.songId:
                raise HTTPException(422, "Session belongs to another song")
            if event.judgment == "undo":
                target = connection.execute("SELECT document FROM practice_events WHERE id=?", (event.supersedes,)).fetchone()
                if not target:
                    raise HTTPException(422, "Undo target not found")
                original = json.loads(target[0])
                if original["songId"] != event.songId or original["lineId"] != event.lineId or original["judgment"] == "undo":
                    raise HTTPException(422, "Invalid undo target")
            conflict = connection.execute("SELECT id FROM practice_events WHERE session_id=? AND sequence=?", (event.sessionId, event.sequence)).fetchone()
            if conflict:
                raise HTTPException(409, "Session sequence already used")
            connection.execute("INSERT INTO practice_events VALUES (?, ?, ?, ?, ?)", (event.id, event.songId, event.sessionId, event.sequence, json.dumps(payload)))
    return {"acknowledged": [event.id for event in batch.events]}


@router.get("/songs/{song_id}/practice-state")
def practice_state(song_id: str) -> dict:
    require_song(song_id)
    with db.connect() as connection:
        rows = connection.execute("SELECT document FROM practice_events WHERE song_id=? ORDER BY rowid", (song_id,)).fetchall()
    events = [json.loads(row[0]) for row in rows]
    return {"events": events}


class SessionPayload(BaseModel):
    id: str
    songId: str
    position: float = Field(ge=0, allow_inf_nan=False)
    mode: Literal["full", "focus"]
    scope: str = ""
    settings: dict = Field(default_factory=dict)
    loop: tuple[int, int] | None = None
    elapsed: float = Field(default=0, ge=0, allow_inf_nan=False)
    updatedAt: str
    plan: dict = Field(default_factory=dict)
    rehearsal: dict = Field(default_factory=dict)


@router.put("/practice-sessions/{session_id}")
def save_session(session_id: str, session: SessionPayload) -> dict:
    require_song(session.songId)
    if session_id != session.id:
        raise HTTPException(422, "Session ID mismatch")
    with db.connect() as connection:
        owner = connection.execute("SELECT song_id FROM practice_sessions WHERE id=?", (session_id,)).fetchone()
        if owner and owner[0] != session.songId:
            raise HTTPException(422, "Session belongs to another song")
        connection.execute("""INSERT INTO practice_sessions VALUES (?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET document=excluded.document, updated_at=excluded.updated_at
            WHERE excluded.updated_at >= practice_sessions.updated_at""",
            (session_id, session.songId, session.model_dump_json(), session.updatedAt))
    return session.model_dump()


@router.get("/songs/{song_id}/session")
def latest_session(song_id: str) -> dict | None:
    require_song(song_id)
    with db.connect() as connection:
        row = connection.execute("SELECT document FROM practice_sessions WHERE song_id=? ORDER BY updated_at DESC LIMIT 1", (song_id,)).fetchone()
    return json.loads(row[0]) if row else None


class BeatSettings(BaseModel):
    bpm: float = Field(gt=20, le=300, allow_inf_nan=False)
    anchor: float = Field(ge=0, allow_inf_nan=False)
    verified: bool = True


@router.get("/songs/{song_id}/beat-map")
def get_beats(song_id: str) -> dict | None:
    require_song(song_id)
    with db.connect() as connection:
        row = connection.execute("SELECT document FROM beat_maps WHERE song_id=?", (song_id,)).fetchone()
    return json.loads(row[0]) if row else None


def store_beats(song_id: str, result: dict) -> dict:
    with db.connect() as connection:
        connection.execute("INSERT INTO beat_maps VALUES (?, ?) ON CONFLICT(song_id) DO UPDATE SET document=excluded.document", (song_id, json.dumps(result)))
    return result


@router.put("/songs/{song_id}/beat-map")
def set_beats(song_id: str, values: BeatSettings) -> dict:
    song, _ = require_song(song_id)
    if values.anchor >= song["duration"]:
        raise HTTPException(422, "Beat anchor must fall within the recording")
    step = 60 / values.bpm
    start = values.anchor - math.floor(values.anchor / step) * step
    beats = [round(start + i * step, 5) for i in range(math.ceil((song["duration"] - start) / step))]
    return store_beats(song_id, dict(bpm=values.bpm, anchor=values.anchor, beats=beats, confidence=1 if values.verified else .5, verified=values.verified, source="manual"))


@router.post("/songs/{song_id}/analyze-beats")
def analyze_beats(song_id: str) -> dict:
    song, paths = require_song(song_id)
    source = paths.get("original") or paths.get("instrumental")
    if not source:
        raise HTTPException(409, "Audio is not ready")
    try:
        from .rhythm import analyze
        result = analyze(source)
    except Exception as exc:
        raise HTTPException(422, f"Beat analysis unavailable: {exc}. Set tempo and a beat anchor manually.") from exc
    return store_beats(song_id, result)


@router.post("/songs/{song_id}/beat-map/confirm")
def confirm_beats(song_id: str) -> dict:
    result = get_beats(song_id)
    if not result:
        raise HTTPException(404, "Analyze beats first")
    return store_beats(song_id, {**result, "verified": True})
